include/boost/math/distributions/dirichlet.hpp

```diff
@@ -0,0 +1,405 @@
+// boost/math/distributions/dirichlet.hpp
+
+// Copyright Mrityunjay Tripathi 2020.
+
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// https://en.wikipedia.org/wiki/Dirichlet_distribution
+// https://mast.queensu.ca/~communications/Papers/msc-jiayu-lin.pdf
+
+// The Dirichlet distribution is a family of continuous multivariate probability
+// distributions parameterized by a vector 'alpha' of positive reals.
+// It is a multivariate generalization of the dirichlet distribution, hence its
+// alternative name of multivariate dirichlet distribution (MBD).
+// Dirichlet distributions are commonly used as prior distributions in
+// Bayesian statistics, and in fact the Dirichlet distribution is the
+// conjugate prior of the categorical distribution and multinomial distribution.
+
+#ifndef BOOST_MATH_DIST_DIRICHLET_HPP
+#define BOOST_MATH_DIST_DIRICHLET_HPP
+
+#include <boost/math/distributions/fwd.hpp>
+#include <boost/math/special_functions/gamma.hpp>
+#include <boost/math/special_functions/digamma.hpp>
+#include <boost/math/distributions/complement.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <boost/math/distributions/detail/common_error_handling.hpp>
+
+#if defined(BOOST_MSVC)
+#pragma warning(push)
+#pragma warning(disable : 4702) // unreachable code
+// in domain_error_imp in error_handling
+#endif
+
+#include <utility>
+
+namespace boost
+{
+namespace math
+{
+namespace dirichlet_detail
+{
+// Common error checking routines for dirichlet distribution function:
+template <class VectorType, class RealType, class Policy>
+inline bool check_concentration(const char *function,
+                                const VectorType &concentration,
+                                RealType *result,
+                                const Policy &pol)
+{
+  for (size_t i = 0; i < concentration.size(); ++i)
+  {
+    if (!(boost::math::isfinite)(i) || (i <= 0))
```

Comment:
`i <=0`? But `i` is a `size_t`. . . 

This will cause warning when used with `Eigen` or any other container that has signed indexes. Use `decltype(concentration.size()) i = 0; i < concentration.size()`. . .

Reply:
BTW, in your unit tests, try to use both `std::vector` and an `Eigen` vector as the template type.

Reply:
It's really silly, it should be ```concentration[i]```.